web-server-lib/web-server/http/request-structs.rkt

```diff
@@ -84,49 +94,47 @@
               #:when (bytes=? f (header-field h)))
     h))
 
-;                                          
-;                                       ;; 
-;                                       ;; 
+;
+;                                       ;;
+;                                       ;;
 ;   ;; ;  ;;    ;;; ; ;; ;;   ;;    ;; ;;;;
-;   ;;;  ;  ;  ;   ;; ;; ;;  ;  ; ;;  ; ;; 
-;   ;;   ;  ;  ;   ;; ;; ;;  ;  ;  ;    ;; 
-;   ;;  ;;;;;;;;   ;; ;; ;; ;;;;;;  ;;  ;; 
-;   ;;   ;     ;   ;; ;; ;;  ;        ;;;; 
-;   ;;   ;     ;   ;;  ; ;;  ;    ;   ;  ; 
+;   ;;;  ;  ;  ;   ;; ;; ;;  ;  ; ;;  ; ;;
+;   ;;   ;  ;  ;   ;; ;; ;;  ;  ;  ;    ;;
+;   ;;  ;;;;;;;;   ;; ;; ;; ;;;;;;  ;;  ;;
+;   ;;   ;     ;   ;; ;; ;;  ;        ;;;;
+;   ;;   ;     ;   ;;  ; ;;  ;    ;   ;  ;
 ;   ;;    ;;;   ;;;;;  ;;;;   ;;;  ;;;   ;;
-;                  ;;                      
-;                  ;;                      
-;                  ;;                      
-;                                          
+;                  ;;
+;                  ;;
+;                  ;;
+;
 
 (define-struct request
-  (method uri headers/raw bindings/raw-promise post-data/raw
-          host-ip host-port client-ip)
+  (method uri headers/raw bindings/raw-promise post-data/raw host-ip host-port client-ip)
   #:transparent)
 
 (define (request-bindings/raw r)
   (force (request-bindings/raw-promise r)))
 
 
-;                                           
-;                                           
-;   ;;     ;;            ;; ;;              
-;   ;;                   ;;                 
+;
+;
+;   ;;     ;;            ;; ;;
+;   ;;                   ;;
 ;   ;;;;   ;; ; ;;;   ;;;;; ;; ; ;;;   ;;;;;
-;   ;;  ;  ;; ;;  ;  ;   ;; ;; ;;  ;  ;  ;  
-;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;;;  ;; 
-;   ;;  ;; ;; ;;  ;;;;   ;; ;; ;;  ;; ;  ;  
-;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;  ;;   
-;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;;;     
-;   ; ;;   ;; ;;  ;;  ;;; ; ;; ;;  ;; ;;;;; 
+;   ;;  ;  ;; ;;  ;  ;   ;; ;; ;;  ;  ;  ;
+;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;;;  ;;
+;   ;;  ;; ;; ;;  ;;;;   ;; ;; ;;  ;; ;  ;
+;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;  ;;
+;   ;;  ;  ;; ;;  ;; ;   ;; ;; ;;  ;;;;
+;   ; ;;   ;; ;;  ;;  ;;; ; ;; ;;  ;; ;;;;;
 ;                                     ;   ;;
-;                                    ;;   ; 
-;                                      ;;;  
-;                                           
+;                                    ;;   ;
+;                                      ;;;
+;
 
 (define-serializable-struct binding
   (id)
-  #:guard (struct-guard/c bytes?)
```

Comment:
Dropping these guards led to (by far) the largest improvement in the whole PR. This is slightly less safe for data that's deserialized, but I think safety is preserved outside this module because I've added `contract-out` forms for these structs (so, even if someone were to deserialize headers with invalid data, this module would be blamed on access to that data).

Reply:
what happens if the guard is `(lambda (x y) (if (bytes? x) x (error 'fail)))`?

Reply:
It's [not bad](https://www.techempower.com/benchmarks/#section=test&shareid=8d0719f4-2571-4931-bf9a-05e6697e95bc&hw=ph&test=composite&a=2) ([diff](https://github.com/racket/web-server/commit/a49d469aaa6cca530daddda19b0efd8b2c4ccde4)), but it doesn't feel great to lose 4k RPS to this. 

My understanding is these structs are serializable so that they can be used in serialized continuations via stuffers and anyone serializing continuations to URLs should be using the HMAC stuffer to prevent tampering (otherwise they probably have much bigger problems), so I don't know if these guards actually buy us much safety. That said, I don't feel too strongly about this and between this and the custom deserializer route, I definitely think this is preferable.

Reply:
I don't think it is worth it